### DISCUSSION TITLE
Using uWS.HttpRequest past its request handler return is forbidden (it is stack allocated).

### DISCUSSION THREAD

**Discussion by zdm** [uNetworking/uWebSockets.js#328] [Q&A] [answered]
Hi,

I want to extend res object with async method, like this:
```
res.sendStrream = async function (path) {...}

// then in http request handler:
uws.get("/", (res) => {
    res.sendStream(path);
})
```
But getting this error at runtime, when calling this new method:
```
(node:44040) UnhandledPromiseRejectionWarning: Using uWS.HttpRequest past its request handler return is forbidden (it is stack allocated).
(
(
```

Is it possible to avoid this error?

**Comment by ghost** [comment, marked as answer]
> Using uWS.HttpRequest past its request handler return is forbidden (it is stack allocated).

This is a message for you to read. When you await anything, that's technically a return. So you cannot use the HttpRequest after this point, you have to copy out whatever you want to keep before you return or await.

**Reply by IonelLupu** [reply]
Does this mean we have to ignore the suggestion from here (https://github.com/uNetworking/uWebSockets.js/blob/master/examples/Headers.js): "not recommended as production solution"?

I have an await too before I use anything from the HttpRequest. Take this handler for example:
```ts
await new Promise(resolve => setTimeout(resolve, 10)) // some async work here
return serverResponse.end(`some header: ${serverRequest.getHeader('content-type')}`)
```

Which means I ignore the suggestion from the example above and do this in production for it to work:
```ts
const headers: Record<string, string> = {}
serverRequest.forEach((name, value) => {
    headers[name] = value
})
await new Promise(resolve => setTimeout(resolve, 10))
return serverResponse.end(`some header: ${headers['content-type']}`)
```
